### PR TITLE
Closes #6845 Rework UI for data clearing and notices for OCI and LRC

### LINKED ISSUE #6845: 3.17 - Data Clearing & Notices
# Description
In 3.17, we will no longer be clearing actions between RUCSS and OCI, so to that fact, we need to change how this was done.

# Scope a solution

## Data Clearing
- In `WP_Rocket\Engine\Saas\Admin\AdminBar`
  - In `add_clean_saas_menu_item`
    - Remove every `atf` check & title with `Clear Critical Images`
    - Rename `Clear Used CSS` to `Clear RUCSS optimizations`
  - In `add_clean_url_menu_item`
    - Remove every `atf` check & title with `Clear Critical Images of this URL`
    - Rename `Clear Used CSS of this URL` to `Clear RUCSS optimizations of this URL`
  - In `display_dashboard_button`
    - Remove every `atf` check
    - Remove every title & label with `Critical Images Cache`
    - Rename `Remove Used CSS Cache` to `Remove Unused CSS`
    - Rename `Clear Used CSS` to `Clear RUCSS optimizations`
  - In File Optimtization > CSS Files > Optimize CSS delivery
    - Rename button `Remove Unused CSS` to `Remove Unused CSS (RUCSS)`

- In `WP_Rocket\Engine\Common\PerformanceHints\Admin`, create similar classes we had in `WP_Rocket\Engine\Saas\Admin` with similar methods exclusive to the performance hints factories and we can also create a class or trait that can be shared by both performance hints and SaaS clearing module as I can see these classes (`AdminBar`, `Clean`, `Notices`) contents can be abstracted that way we don't WRITE EVERYTHING TWICE (WET)
  - Use `Clear Performance Hints data` as text on the WPR admin bar
  - Use Clear performance hints data of this URL as text on WPR admin bar for clearing specific url
  - Use `Performance Hints` as text & `Clear` as label on the Quick Actions on the WPR Dashboard
    - On hover over this button, We will display this text: `This action will clear data for Optimize Critical Images and Lazy Render Content.`

 - In `WP_Rocket\Engine\Common\PerformanceHints\Admin\Subscriber`
   - Replace these events (`rocket_saas_clean_all`, `rocket_saas_clean_url`) with the new ones that will be created in `WP_Rocket\Engine\Common\PerformanceHints\Admin\Clean`

### Notices for data clearing
- In `WP_Rocket\Engine\Common\PerformanceHints\Admin\Controller::truncate_from_admin`
  - Update the notice text - `WP Rocket: Critical images and Lazy Render data was cleared!`

### Notices for upgrade to 3.17
- In `WP_Rocket\Engine\Admin\Beacon\Beacon::get_suggest`
  - Add the new performance hint section and data
- In `WP_Rocket\Engine\Admin\Settings\Page::display_update_notice`
  - Update the condition to 3.17
  - Update the notice text to https://www.notion.so/wpmedia/Lazy-render-content-bf707b5f731a426cb5d1294beb770540?pvs=4#b5d069b2e8d147119a15c5f7408b22b8

- Update Tests

# Acceptance Criteria
- All Cache clearing text in WPR Admin bar & Dashboard Should be updated to those in the [Epic](https://www.notion.so/wpmedia/Lazy-render-content-bf707b5f731a426cb5d1294beb770540?pvs=4#6f496668519c4deba68e36e0ef093574)
- When RUCSS optimization data is cleared, Performance hints data (OCI & LRC) should not be cleared
- When Performance Hints data (OCI & LRC) is cleared RUCSS optimization data should not be cleared
- RUCSS optimization data should be cleared as expected for total & partial clearing, No regression should be introduced
- Performance Hints data (OCI & LRC) should be cleared as expected for total & partial data clearing.
- RUCSS optimization clear button and Performance Hints Optimization clear button should not be displayed when features are disabled both on WPR Admin bar and WPR Dashboard.
- Performance Hints data must be cleared when
  - Switch theme
  - Change permalinks
  - Domain Changed
  - Post Trash
  - Post Delete
  - Post cache clear
  - Comment count update
  - Term Update
  - Term Delete
  - Delete from WPR Admin Bar
  - On WPR update

### PR DESCRIPTION
**Pull request by Khadreal**
# Description

Fixes #6845 

## Documentation
Rework UI for data clearing and notices. Add menu class and traits.

### Technical documentation
Add traits for classes with similar code for admin menu

## Type of change
- [x] Sub-task of epic

## New dependencies
*List any new dependencies that are required for this change.*

## Risks
*List possible performance & security issues or risks, and explain how they have been mitigated.*

# Checklists

## Feature validation

- [x] I validated all the Acceptance Criteria. If possible, provide sreenshots or videos.
- [x] I triggered all changed lines of code at least once without new errors/warnings/notices.
- [x] I implemented built-in tests to cover the new/changed code.

## Code style
- [x] I wrote self-explanatory code about what it does.
- [x] I wrote comments to explain why it does it.
- [x] I named variables and functions explicitely.
- [x] I protected entry points against unexpected inputs.
- [x] I did not introduce unecessary complexity.

### CODE REVIEW

File: inc/Engine/Common/PerformanceHints/ServiceProvider.php [hunk 1]
```diff
@@ -38,6 +44,10 @@ class ServiceProvider extends AbstractServiceProvider {
 		'performance_hints_warmup_queue',
 		'performance_hints_warmup_controller',
 		'performance_hints_warmup_subscriber',
+		'performance_hints_admin_bar',
+		'performance_hints_clean',
+		'performance_hints_notices',
+		'atf_context',
```
Comment (remyperona) [inline inc/Engine/Common/PerformanceHints/ServiceProvider.php]: This service provider is not the one providing the atf_context instance, it doesn't need to be added here.

File: inc/Engine/Common/PerformanceHints/Admin/Clean.php [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	/**
+	 * Truncate performance hints tables when clicking on the dashboard button
+	 *
+	 * @return void
+	 */
+	public function clean_performance_hints(): void {
+		if ( ! isset( $_GET['_wpnonce'] ) || ! wp_verify_nonce( sanitize_key( $_GET['_wpnonce'] ), 'rocket_clean_performance_hints' ) ) {
+			wp_nonce_ays( '' );
+		}
+
```
Comment (remyperona) [inline inc/Engine/Common/PerformanceHints/Admin/Clean.php:20]: we're missing capability check in addition to the nonce check. It was possibly an oversight from the previous implementation, but it's necessary for security hardening.

We can use `rocket_manage_options` capability check.

File: inc/Engine/Common/PerformanceHints/Admin/Clean.php [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+	}
+
+	/**
+	 * Clean performance hints data for current url.
+	 *
+	 * @return void
+	 */
+	public function clean_url_performance_hints(): void {
+		check_admin_referer( 'rocket_clean_performance_hints_url' );
+
```
Comment (remyperona) [inline inc/Engine/Common/PerformanceHints/Admin/Clean.php:44]: Same here

**Review by remyperona — CHANGES_REQUESTED** [review]
Could we see if we can improve the test coverage a bit to pass at least 50%?

**Comment by remyperona** [comment]
Also you can update with the base branch, which will give you access to the LRC context class, to use in addition to the ATF context where needed.

**Comment by MathieuLamiot** [comment]
@Khadreal Have you taken into account @Tabrisrp's comment about test coverage? I can still see the coverage being below 50%. Maybe there is a blocker?

**Review by MathieuLamiot — CHANGES_REQUESTED** [review]
Could you use the apply-filters-typed library instead of apply_filters?
I can see 3 added apply_filters.

**Comment by Khadreal** [comment]
> @Khadreal Have you taken into account @Tabrisrp's comment about test coverage? I can still see the coverage being below 50%. Maybe there is a blocker?

Yes, will add more test

File: inc/Engine/Common/PerformanceHints/Admin/AdminBar.php [hunk 4]
```diff
@@ -0,0 +1,121 @@
...
+	 * @var ContextInterface
+	 */
+	private $atf_context;
+
+	/**
+	 * LRC context.
+	 *
+	 * @var ContextInterface
+	 */
+	private $lrc_context;
```
Comment (jeawhanlee) [inline inc/Engine/Common/PerformanceHints/Admin/AdminBar.php]: We don't need these contexts properties, instead just the factories property.

File: inc/Engine/Common/PerformanceHints/Admin/AdminBar.php [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+
+
+	/**
+	 * Constructor
+	 *
+	 * @param ContextInterface $atf_context ATF context.
+	 * @param ContextInterface $lrc_context LRC context.
+	 * @param string           $template_path Template path.
+	 */
+	public function __construct( ContextInterface $atf_context, ContextInterface $lrc_context, $template_path ) {
```
Comment (jeawhanlee) [inline inc/Engine/Common/PerformanceHints/Admin/AdminBar.php]: We need only to pass the pass the factories here and use their context products.
```suggestion
	public function __construct( array $factories, $template_path ) {
```

File: inc/Engine/Common/PerformanceHints/Admin/AdminBar.php [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+		parent::__construct( $template_path );
+
+		$this->atf_context = $atf_context;
+		$this->lrc_context = $lrc_context;
```
Comment (jeawhanlee) [inline inc/Engine/Common/PerformanceHints/Admin/AdminBar.php]:
```suggestion
		$this->factories = $factories;
```

File: inc/Engine/Common/PerformanceHints/Admin/Controller.php [hunk 7]
```diff
@@ -126,7 +126,7 @@ public function truncate_from_admin( $clean ) {
 			'status'  => 'success',
 			'message' => sprintf(
 				// translators: %1$s = plugin name.
-				__( '%1$s: Critical images cleared!', 'rocket' ),
+				__( '%1$s: WP Rocket: Critical images and Lazy Render data was cleared!', 'rocket' ),
```
Comment (jeawhanlee) [inline inc/Engine/Common/PerformanceHints/Admin/Controller.php]: We are already passing the formatted string in the first parameter in `sprintf`
```suggestion
				__( '%1$s: Critical images and Lazy Render data was cleared!', 'rocket' ),
```

File: inc/Engine/Common/PerformanceHints/Admin/Notices.php
Comment (jeawhanlee) [inline inc/Engine/Common/PerformanceHints/Admin/Notices.php:1]: Pass the factories as dependency instead.

File: inc/Engine/Common/PerformanceHints/ServiceProvider.php [hunk 8]
```diff
@@ -95,10 +105,24 @@ public function register(): void {
 				]
 			);
 
+		$this->getContainer()->add( 'performance_hints_notices', Notices::class )
+			->addArgument( $this->getContainer()->get( 'atf_context' ) )
+			->addArgument( $this->getContainer()->get( 'lrc_context' ) );
```
Comment (jeawhanlee) [inline inc/Engine/Common/PerformanceHints/ServiceProvider.php]: Pass the factories as dependency.

File: inc/Engine/Common/PerformanceHints/ServiceProvider.php [hunk 9]
```diff
@@ -95,10 +105,24 @@ public function register(): void {
...
+
+		$this->getContainer()->add( 'performance_hints_admin_bar', Adminbar::class )
+			->addArgument( $this->getContainer()->get( 'atf_context' ) )
+			->addArgument( $this->getContainer()->get( 'lrc_context' ) )
```
Comment (jeawhanlee) [inline inc/Engine/Common/PerformanceHints/ServiceProvider.php]: Pass the factories as dependency

File: inc/Engine/Saas/Admin/Clean.php [hunk 10]
```diff
@@ -25,31 +28,7 @@ public function clean_saas() {
 		 */
 		$clean = apply_filters( 'rocket_saas_clean_all', [] );
```
Comment (jeawhanlee) [inline inc/Engine/Saas/Admin/Clean.php]:
```suggestion
		$clean = wpm_apply_filters_typed( 'array', 'rocket_saas_clean_all', [] )
```

**Review by jeawhanlee — CHANGES_REQUESTED** [review]

**Review by jeawhanlee — APPROVED** [review]

**Review by MathieuLamiot — APPROVED** [review]

**Comment by hanna-meda** [comment]
Thank you all for the work on this one. Have a couple of minor changes on UI:
1. Under WPR Dashboard > Quick Actions > Performance Hints: Spacing on the `Clear` button should be consistent with the ones above
Screenshot 2024-08-19 at 12 52 39
2. On View page of any published URL, the `Performance Hints` text should start in upper case
Screenshot 2024-08-19 at 12 41 24

**Comment by Khadreal** [comment]
@hanna-meda the modified UI
Screenshot 2024-08-19 at 13 11 32
Screenshot 2024-08-19 at 13 11 42

**Comment by hanna-meda** [comment]
@Khadreal, "Clear" button still looks the same to me. Checked on Chrome & FF (Mac) with cleared site cache & cookies:
Screenshot 2024-08-19 at 16 02 06

**Comment by hanna-meda** [comment]
@Khadreal, using filters
`add_filter( 'rocket_above_the_fold_optimization', function( $flag ){ return false; } );`
`add_filter( 'rocket_lazy_render_content_optimization', function( $flag ){ return false; } );`
 to disable LRC & ATF does not seem to work. We will still see `Clear Performance Hints Data` in Admin bar and `Clear` button under the Quick Actions from Dashboard > Performance Hints section. Is this dependent on #6828 or do we have an issue on PR? If you can have a look. Thank you.

**Comment by Khadreal** [comment]
@hanna-meda buttons look good from my end using diff browser, could you confirm you're using latest branch.
FF
Screenshot 2024-08-19 at 16 56 13
Safari
Screenshot 2024-08-19 at 16 56 13
Chrome
Screenshot 2024-08-19 at 16 57 43


Filter 
This should work for now `add_filter( 'rocket_above_the_fold_optimization', function( $flag ){ return false; } );`, `add_filter( 'rocket_lazy_render_content_optimization', function( $flag ){ return false; } )` this is not fully working as it's dependent on other components

**Review by jeawhanlee — APPROVED** [review]

**Comment by jeawhanlee** [comment]
Taking this back to in progress in respect to @benorfaz comment [here](https://github.com/wp-media/wp-rocket/issues/6845#issuecomment-2298143530)

File: inc/Engine/Admin/Settings/Page.php [hunk 11]
```diff
@@ -2248,7 +2248,7 @@ public function display_update_notice() {
 				'dismissible'    => '',
 				'message'        => sprintf(
 					// translators: %1$s: opening strong tag, %2$s: closing strong tag, %3$s: opening a tag, %4$s: opening a tag.
-					__( '%1$sWP Rocket:%2$s the plugin has been updated to the 3.17 version. Our brand new feature %3$sLazy Render Content%4$s. Check out our documentation to learn more about it.', 'rocket' ),
+					__( '%1$sWP Rocket:%2$s the plugin has been updated to the 3.17 version. New feature: %3$sLazy Render Content%4$s. Check out our documentation to learn more about it.', 'rocket' ),
```
Comment (jeawhanlee) [inline inc/Engine/Admin/Settings/Page.php:2251]: Did product influence the change of this text?

Comment (Khadreal) [reply inc/Engine/Admin/Settings/Page.php:2251]: Yes

Comment (hanna-meda) [reply inc/Engine/Admin/Settings/Page.php:2251]: @jeawhanlee It was I who pointed out given text from [Notion](https://www.notion.so/wpmedia/Lazy-render-content-bf707b5f731a426cb5d1294beb770540?pvs=4#b5d069b2e8d147119a15c5f7408b22b8)

Comment (jeawhanlee) [reply inc/Engine/Admin/Settings/Page.php:2251]: Thanks @hanna-meda I asked @wp-media/product a question [here](https://wp-media.slack.com/archives/CUT7FLHF1/p1724152527252139) though

Comment (jeawhanlee) [reply inc/Engine/Admin/Settings/Page.php:2251]: Ok by product.

**Review by jeawhanlee — APPROVED** [review]

**Review by hanna-meda — APPROVED** [review]
Related Test Report below:
[testrail-report-632.pdf](https://github.com/user-attachments/files/16692187/testrail-report-632.pdf)
with 3 tests failing:

- 1 regression, new ticket #6883
- 1 existing ticket #6704
- 1 failing also on trunk, ~still investigating~ new ticket #6885 

Blocked TCs will be retested when we'll be able to check the new LRC table from DB.

**Comment by MathieuLamiot** [comment]
Merged as per QA request at the daily.

**Pull request merged**